### ISSUE TITLE
[suggestion] NUnit1028 could be improved

### ISSUE THREAD

**Issue by Bartleby2718** [nunit/nunit.analyzers#714]
Problem 1:
https://github.com/nunit/nunit.analyzers/blob/e52e3a1696c43b10e1a76b9e4e9309646c1fe808/src/nunit.analyzers/NonTestMethodAccessibilityLevel/NonTestMethodAccessibilityLevelAnalyzer.cs#L63-L75
clearly says that NUnit1028 detects public _or internal_ methods, but the title, message, and description don't mention `internal`:
https://github.com/nunit/nunit.analyzers/blob/e52e3a1696c43b10e1a76b9e4e9309646c1fe808/src/nunit.analyzers/NonTestMethodAccessibilityLevel/NonTestMethodAccessibilityLevelConstants.cs#L5-L7

Problem 2:
https://github.com/nunit/nunit.analyzers/blob/e52e3a1696c43b10e1a76b9e4e9309646c1fe808/src/nunit.analyzers/NonTestMethodAccessibilityLevel/NonTestMethodAccessibilityLevelCodeFix.cs#L61-L101 makes the offending method `private`, but it should suggest so only if it isn't used anywhere else. Not sure if this can be done, but it should definitely be possible to reduce false positives.

**Comment by manfred-brands** [comment]
@Bartleby2718 Yes, the wording could be improved, to something like `The non-test method is not private`.
Unfortunately there is no way to see if the method is used elsewhere. The analyzer pass only knows about the current "file", not about others. The code fix is to help, not blindly follow. If method are really used it would involve moving the method to another helper class or so.

**Comment by marcusbrunello** [comment]
I am confused about this. Is it a good choice to make them internal?? NUnit1028 is complaining about them if internal ALSO.

**Comment by manfred-brands** [comment]
No. A test should be self-contained not expose methods to be called from elsewhere. Depending on what those methods do they could interfere with a test in progress and affect the result.
The only way to be certain is to only have private non-test methods.

If a method is shared between tests then they should be moved to a separate non-test class.

**Comment by marcusbrunello** [comment]
I see... and then how should those non-test classes be annotated?
Can you check/do you know if I have to annotate the last class in these group?
```c#
[SetUpFixture]
public partial class Testing
{
    private static ITestDatabase s_database;
    private static CustomWebApplicationFactory s_factory = null!;
    private static IServiceScopeFactory s_scopeFactory = null!;
    private static string? s_userId;

    [OneTimeSetUp]
    public async Task RunBeforeAnyTests()
    {
        s_database = await TestDatabaseFactory.CreateAsync();

        s_factory = new CustomWebApplicationFactory(s_database.GetConnection());

        s_scopeFactory = s_factory.Services.GetRequiredService<IServiceScopeFactory>();
    }

    internal static async Task<TResponse> SendAsync<TResponse>(IRequest<TResponse> request)
    {
        using var scope = s_scopeFactory.CreateScope();

        var mediator = scope.ServiceProvider.GetRequiredService<ISender>();

        return await mediator.Send(request);
    }
```

**Comment by manfred-brands** [comment]
For your case, I would removal all attributes from your class and create a new SetUpFixture class which calls `Testing.RunBeforeAnyTests()` in its `OneTimeSetUp` method.

Your test code can then keep calling `Testing.SendAsync` as before.

**Comment by Greybird** [comment]
I also have a use case where a method comes from a static interface definition, which mandates the public visibility. I believe this should be ignored by the rule.
I have a base class for fixtures, which allows me to force the definition of a method to provide test cases:
```
public interface IInvalidPayloadValidators
{
    static abstract IEnumerable<TestCaseData> InvalidPayloads();
}
```
As I cannot make implementation non public, the NUnit1028 violation is noise.

**Comment by manfred-brands** [comment]
@Greybird I will take your comment into account.

In the mean time you can make an explicit implementation:

```csharp
public class SomeFixture :  IInvalidPayloadValidators
{
    static IEnumerable<TestCaseData> IInvalidPayloadValidators.InvalidPayloads()
    {
        return Array.Empty<TeatCaseData>();
    }
}
```
